File: ProblemSet/464 · 整数排序 II/python/464 · 整数排序 II 归并排序 非递归实现.py
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        self.n = len(A)
        self.mergeSort(A)

    def mergeSort(self, nums):
        tmp = [None] * (self.n)
        subLen = 1
        while subLen < self.n:
            self.mergePass(nums, tmp, subLen)
            subLen <<= 1
            self.mergePass(tmp, nums, subLen)
            subLen <<= 1

    def mergePass(self, nums, tmp, subLen):
        pos = 0
        while pos <= self.n - 2 * subLen:
            self.merge(nums, tmp, pos, pos + subLen - 1, pos + subLen, pos + 2 * subLen - 1)
            pos += 2 * subLen
        if pos + subLen < self.n:
            self.merge(nums, tmp, pos, pos + subLen - 1, pos + subLen, self.n - 1)
        else:
            tmp[pos : self.n] = nums[pos : self.n]

    def merge(self, nums, tmp, lStart, lEnd, rStart, rEnd):
        i, j, pos = lStart, rStart, lStart
        while i <= lEnd and j <= rEnd:
            if nums[i] <= nums[j]:
                tmp[pos] = nums[i]
                i += 1
            else:
                tmp[pos] = nums[j]
                j += 1
            pos += 1
        tmp[pos : rEnd + 1] = nums[i : lEnd + 1] if i <= lEnd else nums[j : rEnd + 1]

File: ProblemSet/464 · 整数排序 II/java/464 · 整数排序 II 归并排序 非递归实现.java
public class Solution {
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        mergeSort(A);
    }
    
    private void mergeSort(int[] nums) {
        int[] tmp = new int[nums.length];
        int subLen = 1;
        while (subLen < nums.length) {
            mergePass(nums, tmp, subLen);
            subLen <<= 1;
            mergePass(tmp, nums, subLen);
            subLen <<= 1;
        }
    }

    private void mergePass(int[] nums, int[] tmp, int subLen) {
        int pos = 0;
        while (pos <= nums.length - 2 * subLen) {
            merge(nums, tmp, pos, pos + subLen - 1, pos + subLen, pos + 2 * subLen - 1);
            pos += 2 * subLen;
        }
        if (pos + subLen < nums.length) merge(nums, tmp, pos, pos + subLen - 1, pos + subLen, nums.length - 1);
        else {
            for (int i = pos; i < nums.length; ++i)
                tmp[i] = nums[i];
        }
    }

    private void merge(int[] nums, int[] tmp, int lStart, int lEnd, int rStart, int rEnd) {
        int i = lStart, j = rStart, pos = lStart;
        while (i <= lEnd && j <= rEnd) tmp[pos++] = nums[i] <= nums[j] ? nums[i++] : nums[j++];
        while (i <= lEnd) tmp[pos++] = nums[i++];
        while (j <= rEnd) tmp[pos++] = nums[j++];
    }
}